### ISSUE TITLE
How is feature importance calculated for XGBoost Regressor in python?

### ISSUE THREAD

**Issue by hmodea** [dmlc/xgboost#3216]
Not sure this is the write place to ask but I'm really stuck in this as my trying to compare both Random forests and XGBoost regressors in terms of features importances for a project. In case of classification classification it's the F1 score but as far as I understand that makes no sense in regression as we don't have the notion of precision or recall

**Comment by dangoldner** [comment]
@hmodea Does [this](https://stackoverflow.com/questions/34218245/how-is-the-feature-score-importance-in-the-xgboost-package-calculated) help?

**Comment by hmodea** [comment]
Yes ! I actually confused F-score with F1-score which is only relevant for Classification tasks. Thanks a lot!

**Comment by ConorMcNamara** [comment]
To add with @dangoldner xgboost actually has three ways of calculating feature importance.

From the [Python docs](https://xgboost.readthedocs.io/en/latest/python/python_api.html) under class 'Booster': 

> ‘weight’ - the number of times a feature is used to split the data across all trees. ‘gain’ - the average gain of the feature when it is used in trees. ‘cover’ - the average coverage of the feature when it is used in trees

Currently, calling get_fscore() returns 'weight' while calling feature_importances_() returns weight divided by the sum of all feature weights.

**Comment by dangoldner** [comment]
'gain' is [defined in the docs](http://xgboost.readthedocs.io/en/latest/model.html), but the word 'coverage' does not appear in the docs anywhere (apart from the feature importance docs pointed to by @ConorMcNamara). I'm not sure what it means. Is there a concise definition of coverage we could link to or add to the docs?

**Comment by ConorMcNamara** [comment]
@dangoldner There's [this post on Stack Exchange](https://datascience.stackexchange.com/questions/12318/how-do-i-interpret-the-output-of-xgboost-importance) that gives ELI5 definitions of gain, weight and cover. 

Even then, cover seems the most difficult to understand as well as the least important in terms of measuring feature importance.

**Comment by dangoldner** [comment]
@ConorMcNamara - thanks. There does seem to be room for improvement there. 

Not sure if the xgboost docs need to educate on boosting vocabulary, or assume it. If the latter, this can probably be closed.
